opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -208,201 +208,159 @@ void extendGIDtoRank(const GraphOfGrid<Dune::CpGrid>& gog,
     if (cc.rank()==root)
     {
-        for (int i=0; i<cc.size(); ++i)
+        assert((int)exportedCells.size()==cc.size());
+        assert((int)exportedCells[root].size()==0);
+        // check send success for each well
+        std::vector<MPI_Request> requestSize(2*(cc.size()-1));
+        std::vector<int> nrWells(cc.size()-1); // store well sizes (can not send temporary)
+
+        for (int i=0; i<cc.size()-1; ++i)
         {
-            if (i!=root)
-            {
-                int numWells = exportedWells[i].size();
-                int totsize = numWells+1;
-                for (const auto& well : exportedWells[i])
-                {
-                    totsize += well.size();
-                }
-                // data: {N, size0, data0, size1, data1,... size(N-1), data(N-1)},
-                std::vector<int> commData;
-                commData.reserve(totsize);
-                commData.push_back(numWells);
-                for (const auto& well : exportedWells[i])
-                {
-                    commData.push_back(well.size());
-                    for (const auto& gID : well)
-                    {
-                        commData.push_back(gID);
-                    }
-                }
-                assert(totsize==(int)commData.size());
-                int tag = 37; // a random number
-                MPI_Send(&totsize, 1, MPI_INT, i, tag++, cc);
-                MPI_Send(commData.data(), totsize, MPI_INT, i, tag, cc);
-            }
+            int ii=i+(int)(i>=root); // ii takes values {0,...,mpisize-1} but skips root
+            int tag = 37; // a random number
+            nrWells[i]=exportedCells[ii].size();
+            MPI_Isend(&nrWells[i], 1, MPI_INT, ii, tag++, cc, &requestSize[2*i]);
+            MPI_Isend(exportedCells[ii].data(), nrWells[i], MPI_INT, ii, tag++, cc, &requestSize[2*i+1]);
         }
+        // wait till all messages are completed
+        MPI_Waitall(requestSize.size(), requestSize.data(), MPI_STATUS_IGNORE);
     }
     else // receive data from root
     {
         int tag = 37; // a random number
-        int totsize;
-        MPI_Recv(&totsize, 1, MPI_INT, root, tag++, cc, MPI_STATUS_IGNORE);
-        std::vector<int> receivedData(totsize);
-        MPI_Recv(receivedData.data(), totsize, MPI_INT, root, tag, cc, MPI_STATUS_IGNORE);
-
-        int numWells = receivedData[0];
-        result.resize(numWells);
-        int index = 1;
-        for (int i=0; i<numWells; ++i)
-        {
-            int wellSize = receivedData[index++];
-            assert(index+wellSize<=totsize);
-            const auto dataBegin = receivedData.begin()+index;
-            result[i] = std::vector<int>(dataBegin, dataBegin+wellSize);
-            index+=wellSize;
-        }
+        int nrWells;
+        MPI_Recv(&nrWells, 1, MPI_INT, root, tag++, cc, MPI_STATUS_IGNORE);
+        result.resize(nrWells);
+        MPI_Recv(result.data(), nrWells, MPI_INT, root, tag++, cc, MPI_STATUS_IGNORE);
```

Comment:
Same nitpicking, sorry:
```suggestion
        MPI_Recv(&nrWells, 1, MPI_INT, root, tag, cc, MPI_STATUS_IGNORE);
        result.resize(nrWells);
        MPI_Recv(result.data(), nrWells, MPI_INT, root, tag + 1, cc, MPI_STATUS_IGNORE);
```